java/src/main/java/org/whispersystems/libsignal/util/Base32.java

```diff
@@ -92,66 +67,47 @@ public String encodeToString(final byte[] bytes) {
      * @return Decoded <code>base32</code> String as a raw byte array.
      */
     public byte[] decodeFromString(final String base32) {
-        int i, index, lookup, offset, digit, replacement;
-        byte[] bytes = new byte[base32.length() * 5 / 8];
-
-        for (i = 0, index = 0, offset = 0; i < base32.length(); i++) {
-            lookup = base32.charAt(i);
-
-            /* Guard against unrecognized characters */
-            if (lookup < 0 || lookup >= base32Lookup.length) {
-                throw new InvalidCharacterException((char) lookup);
-            }
-
-            /* Replace characters using the replacement table */
-            replacement = replacementLookup[lookup];
-            if (replacement != 0xFF) {
-                lookup = replacement;
-            }
-
-            digit = base32Lookup[lookup];
-
-            /* If this digit is not in the table, throw an exception */
-            if (digit == 0xFF) {
-                throw new InvalidCharacterException((char) lookup);
-            }
-
-            if (index <= 3) {
-                index = (index + 5) % 8;
-                if (index == 0) {
-                    bytes[offset] |= digit;
-                    offset++;
-                    if (offset >= bytes.length)
-                        break;
-                } else {
-                    bytes[offset] |= digit << (8 - index);
-                }
-            } else {
-                index = (index + 5) % 8;
-                bytes[offset] |= (digit >>> index);
-                offset++;
-
-                if (offset >= bytes.length) {
-                    break;
-                }
-                bytes[offset] |= digit << (8 - index);
+        int encodedLength = base32.length();
+        int paddedLength = (int) Math.ceil(((double) encodedLength) / 8.0) * 8;
+        final StringBuilder base32Builder = new StringBuilder(paddedLength);
+        for (char c : base32.toCharArray()) {
+            // This is not constant time, but it only applies if people have manually fat-fingered
+            // a character.
```

Comment:
It would be weird and maybe confusing to read in the source... but I suppose you could make it constant time by having the mapping also include the rest of the characters mapped to themselves?